### ISSUE TITLE
Optional Outputs

### ISSUE THREAD

**Issue by cjllanwarne** [broadinstitute/cromwell#1838]
Support this:
```
task foo {
  command { ... }
  output {
    File? optionalOutput = "am_I_even_here.txt"
}
```

Bonus points? (actually I don't think I like this):
```
task foo {
  command { ... }
  output {
    String? optionalOutput = read_string("am_I_even_here.txt")
}
```

**Comment by katevoss** [comment]
@cjllanwarne has this been implemented? If not, can you give me more information about when a user would want this?

**Comment by patmagee** [comment]
Ive also come across a few scenarios where a task was run that produced different output, depending on whether or not some flag was set in an input file.

In most cases I just name the files the same, or use a glob, but there have been some scenarios where the resulting names were totally different with different extensions and I had to create seperate tasks for them.

**Comment by cowmoo** [comment]
I want to +1 on this issue. 

This is useful in a WDL as you can add logic to make the WDL easier/cleaner to read and maintain:

>  if (defined(upstream_task.optional_output)) {...}

**Comment by dheiman** [comment]
@katevoss we've been asking for this for nearly 2 years (see [Optional Outputs #20](https://github.com/openwdl/wdl/issues/20)) .

We currently use work-arounds involving a "null" file that we keep in an external bucket, and checking if an optional input that would result in an optional output is defined:
`opt_out_file = "${if defined(opt_in) then 'opt_out.txt' else null_file}"`

We use this work-around so often that it is in the WDL template we use to create new workflows. The more complex the workflow, the more complex this work-around becomes, often involving boolean flags be passed alongside the optional inputs.

Our WDLs would be far cleaner with some form of optional output syntax.

**Comment by geoffjentry** [comment]
Hi @dheiman - I know you're already aware as you're on the openwdl mailing list but just wanted to point out that w/ the move to the OpenWDL governance Cromwell issues aren't the path to effect change in WDL. 

I've been looking at how to transfer WDL related issues from this repo to that one but in the meantime note that the way to have changes find their way into the spec would be to discuss the topic on the mail list/gitter and ultimately to open a PR with proposed changes to the spec. See the [rfc protocol](https://github.com/openwdl/wdl/blob/master/GOVERNANCE.md#rfc-process) for more details.

**Comment by geoffjentry** [comment]
@dheiman These are all true statements. My point was just filing an issue is no guarantee that anyone will do the work of defining the spec changes and that you have the ability to make it happen now. In effect, what I'm saying is you have *more* ability now than you did before in that one can continue to wait until someone gets around to it (which was always the case) or one can put forth the effort to make it happen themselves (which is new)

**Comment by noblem** [comment]
Thanks for that additional info, Jeff.  Just so I can be clear: other than "age" and "more detail," is there an essential difference between the aims of what David linked to this thread and the aims of the OP?  The goal of the OP is to have the WDL spec change to support optional outputs, correct?  And David is essentially adding a +1 [or +10, if that's legal :) ] to this.  Regardless, we're happy to put forth effort towards the OpenWDL forums as you suggest, and possibly even contribute code when possible.

**Comment by geoffjentry** [comment]
@noblem The primary difference is that the path forward to have the spec change is to present a PR and not to have an issue open. My point in my last reply is that someone might see it at some point and make a PR for you but there's no guarantee of that happening in a timely fashion whereas opening a PR against the spec yourself does guarantee that happening :)

FWIW I realize that the switcharoo in process can be frustrating. As an example of how we feel the pain as well, there are cases where we (Cromwell team) forgot to update the spec document when we made changes which now means that those changes aren't actually part of the spec. So for instance [this PR](https://github.com/openwdl/wdl/pull/148) is just trying to specify behavior Cromwell already does but is likely to wind up different than what we did & thus making Cromwell non-conforming. That's what we get for being lazy 6 months ago :)
